compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -24,17 +25,276 @@ use rustc_span::Span;
+    }
+    .to_string();
+
+    // add outer_fn name to ad_name to make it unique, in case users apply autodiff to multiple
+    // functions. Unwrap will only panic, if LLVM gave us an invalid string.
+    let name = llvm::get_value_name(outer_fn);
+    let outer_fn_name = std::ffi::CStr::from_bytes_with_nul(name).unwrap().to_str().unwrap();
+    ad_name.push_str(outer_fn_name.to_string().as_str());
+
+    // Let us assume the user wrote the following function square:
+    //
+    // ```llvm
+    // define double @square(double %x) {
+    // entry:
+    //  %0 = fmul double %x, %x
+    //  ret double %0
+    // }
+    // ```
+    //
+    // The user now applies autodiff to the function square, in which case fn_to_diff will be `square`.
+    // Our macro generates the following placeholder code (slightly simplified):
+    //
+    // ```llvm
+    // define double @dsquare(double %x) {
+    //  ; placeholder code
+    //  return 0.0;
+    // }
+    // ```
+    //
+    // so our `outer_fn` will be `dsquare`. The unsafe code section below now removes the placeholder
+    // code and inserts an autodiff call. We also add a declaration for the __enzyme_autodiff call.
+    // Again, the arguments to all functions are slightly simplified.
+    // ```llvm
+    // declare double @__enzyme_autodiff_square(...)
+    //
+    // define double @dsquare(double %x) {
+    // entry:
+    //   %0 = tail call double (...) @__enzyme_autodiff_square(double (double)* nonnull @square, double %x)
+    //   ret double %0
+    // }
+    // ```
+    unsafe {
+        // On LLVM-IR, we can luckily declare __enzyme_ functions without specifying the input
+        // arguments. We do however need to declare them with their correct return type.
+        // We already figured the correct return type out in our frontend, when generating the outer_fn,
+        // so we can now just go ahead and use that. FIXME(ZuseZ4): This doesn't handle sret yet.
+        let fn_ty = llvm::LLVMGlobalGetValueType(outer_fn);
+        let ret_ty = llvm::LLVMGetReturnType(fn_ty);
+
+        // LLVM can figure out the input types on it's own, so we take a shortcut here.
+        let enzyme_ty = llvm::LLVMFunctionType(ret_ty, ptr::null(), 0, True);
+        let ad_fn = llvm::LLVMRustGetOrInsertFunction(
+            llmod,
+            ad_name.as_ptr() as *const c_char,
+            ad_name.len().try_into().unwrap(),
+            enzyme_ty,
+        );
+        // Otherwise LLVM might inline our temporary code before the enzyme pass has a chance to
+        // do it's work.
+        let attr = llvm::AttributeKind::NoInline.create_attr(llcx);
+        attributes::apply_to_llfn(ad_fn, Function, &[attr]);
+
+        // first, remove all calls from fnc
+        let entry = llvm::LLVMGetFirstBasicBlock(outer_fn);
+        let br = llvm::LLVMRustGetTerminator(entry);
+        llvm::LLVMRustEraseInstFromParent(br);
+
+        let builder = llvm::LLVMCreateBuilderInContext(llcx);
```

Comment:
I think this should be using the `Builder::with_cx` function and use methods on the `Builder` instead of accessing the raw llvm builder directly for everything

Reply:
I've been trying this for a bit, but right now I use a `&CodegenContext` when defining the `autodiff` function in `compiler/rustc_codegen_ssa/src/traits/write.rs` under the `WriteBackendMethods` trait. 

The autodiff implementation takes a `LlvmCodegenBackend`, so I have the llcx, but not the different `CodegenCx` from the Builder, defined in `compiler/rustc_codegen_llvm/src/context.rs`. I might miss a way to get the CodegenCx, but right now I don't see how I can move the code over, while keeping autodiff under the `WriteBackendMethods` trait?

Reply:
Also for context, we *must* run autodiff on the single fat-lto module (for the foreseeable future, though there are efforts to remove it over the next months/years) and the CodegenCx seems more like something more low-level, per individual CGU, so it wouldn't be the right level, despite otherwise having nice helpers?

Reply:
Yea this isn't possible without refactoring the builder/codegen_cx quite a bit. It's indeed for one cgu, though it doesn't seem too bad to generate a dummy cgu (it's just a rust struct with no LLVM information in it) for now and use https://github.com/rust-lang/rust/blob/904d8f6b39ce98fb890ac3381a710f071ddae44d/compiler/rustc_codegen_llvm/src/context.rs#L474 

It gives autodiff strictly more powers than it needs to (e.g. all the caches are useless for it), but then we at least are using the existing helpers and can then analyze what to refactor.

@nikic what do you think?

Reply:
I moved it all over, but not marking as resolved so nikita can see it.